### DISCUSSION TITLE
Nextcloud OIDC Login "is_admin" Attribute

### DISCUSSION THREAD

**Discussion by Smithoo4** [authelia/authelia#6010] [General]
I have been playing around with Authelia and Nextcloud. Following the guide [https://www.authelia.com/integration/openid-connect/nextcloud/](url) I have got things working. When I was reviewing the Nextcloud OIDC app documentation [https://github.com/pulsejet/nextcloud-oidc-login,](url) I found there is a log in attribute "is_admin". From what I gather, this is a binary ODIC attribute that if set to true for a user, that user would be an admin in Nextcloud. Is there a way for Authelia to support this, as it is a feature I would like to use.

I am currently using the file authentication_backend, but have looked at something like LDAP using LLDAP. Not sure if that will play any part in this.

Thank you

**Comment by james-d-elliott** [comment]
There is no current way and no clear way to do this that's going to be work for everyone without considerable work on our end.

What attribute in LLDAP that is truthy would represent the fact they are an admin in NextCloud? Wouldn't group membership (i.e. a group named `NextCloud Admns`) determining who is and is not an admin be a more appropriate thing for NextCloud (via the plugin or other means) to support?

**Comment by hendrik1120** [comment]
I would suggest using both OIDC and LDAP with your nextcloud instance. The Nextcloud LDAP backend will manage the users and groups, while OIDC will just handle login (user creation disabled). This will provide the feature you need and somewhat enhance security since the user needs to be present in both backends. In the nextcloud ldap backend, you can set which groups are allowed to use admin priviledges.
You can use LLDAP for both authelia and nextcloud. It works really well despite still being in very active development.

My personal opinion on this is: Unless you absolutely need to, just don't set nextcloud admin privileges automatically. They are way too powerful and really only one tightly controlled account should have them, if at all. You can manage your whole nextcloud instance through the cli.

**Comment by Smithoo4** [comment]
Thank you for the response and advice.

I have investigated LDAP and Nextcloud a little more. On the Nextcloud side it looks like you are currently not able to have an LDAP group that would map the Nextcloud admin group: https://github.com/nextcloud/server/issues/6428.

So far the closest thing to be able to do this I have come across is the ‘is_admin’ oidc_login_attributes in the Nextcloud OIDC Login app; https://github.com/pulsejet/nextcloud-oidc-login. However, this seems to be more geared to using something like Keycloak which I am not sure I want to get into. Also, I can fully appreciate implementing support for something like this in Authelia is probably not worth the effort.

The advice to not set admin privilege and just manage Nextcloud through the cli seem to be pretty good.

 Thank you for your time

**Reply by hendrik1120** [reply]
Oh, I am sorry, I really thought this basic LDAP feature is supported by Nextcloud. I did my setup quite a while ago.

There is an official Nextcloud OIDC backend now. I just tested it, and it works great with authelia, but it also doesn't allow setting the admin group. Since Authelia currently doesn't support access control for OpenID connect, I still have to use the LDAP backend to limit access to nextcloud.

**Reply by james-d-elliott** [reply]
> Also, I can fully appreciate implementing support for something like this in Authelia is probably not worth the effort.

It is something that is likely to be implemented at some stage. Probably with or after #2868,

My only issue with it in general is that realistically providers provide information known about a user, which typically includes a list of groups, a role, or a list of roles. Why wouldn't they want to give users a way to check that instead?

**Reply by Smithoo4** [reply]
@hendrik1120 not sure if this what you are referring or not, but you can use the Nextcloud OIDC backend with Authelia to limit access to Nextcloud to a group, [https://github.com/pulsejet/nextcloud-oidc-login#login-filter](url). This is what I did and everything seem to work.

1.  Create a nextcloud group and add user to it. I did this in my users_database.yml, but I would assume it would work in the same if you are using an LDAP back end to Authelia, I just did not test it.
2.    All the following to your Nextcloud config.php:

```
  'oidc_login_attributes' => array (
      'id' => 'preferred_username',
      'name' => 'name',
      'mail' => 'email',
      'groups' => 'groups',
      'login_filter' => 'groups',
  ),
  'oidc_login_filter_allowed_values' => array('nextcloud'),
```

This limit access to uses in the nextcloud group.

Cheers

**Reply by hendrik1120** [reply]
@Smithoo4 There are currently three OIDC implementations for Nextcloud that I know of.
Yes, the community plugin pulsejet/nextcloud-iodc-login kinda has support for this. It was changed multiple times in the beginning, so I just used LDAP.

The official plugin nextcloud/user_oidc doesn't have support for this at all. That usually isn't a problem since the IDP can restrict access. This plugin is also within the Nextcloud hackerone scope, I would consider it being more tested and more secure. This really is just personal preference. Use whatever fits your needs most.

**Comment by Smithoo4** [comment]
@hendrik1120 if you don't mind me asking how did you get the nextcloud/user_oidc to work with the Nextcloud LDAP backend? I have been trying on and off again to get this to work, with no such luck. I can get them both of them to work independently, but they just create their own users. Or I get an "Internal Server Error" when I log in with Authelia after turn off auto provisioning in user_oidc.

I think the issue has to do with the "ldapExpertUsernameAttr" setting in the user_ldap app and the "mapping-uid" settings in user_oidc and getting them to match. I have tried and few different things here. However, I don't seem to be smart enough to find the right ones that will work.  I thought uid and sub respectively would have been correct.

Below are the steps I have taken:

```
# Setup LDAP
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ app:install user_ldap
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ app:enable user_ldap
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:create-empty-config

# EDIT: domain
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapHost "lldap"
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapPort 3890
# EDIT: admin user
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapAgentName "uid=ro_admin,ou=people,dc=example, dc=com"
# EDIT: password
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapAgentPassword "Password"
# EDIT: Base DN
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapBase "dc=example, dc=com"
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapBaseUsers "dc=example, dc=com"
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapBaseGroups "dc=example, dc=com"
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapConfigurationActive 1
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapLoginFilter "(&(objectclass=person)(uid=%uid))"
# EDIT: nextcloud_users group, contains the users who can login to Nextcloud
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUserFilter "(&(objectclass=person)(memberOf=cn=nextcloud_users,ou=groups,dc=example, dc=com))"
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUserFilterMode 0
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUserFilterObjectclass person
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 turnOnPasswordChange 0
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapCacheTTL 600
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapExperiencedAdmin 0
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapGidNumber gidNumber
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapExpertUsernameAttr uid
# EDIT: list of application groups
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapGroupFilter "(&(objectclass=groupOfUniqueNames)(|(cn=friends)(cn=family)))"
# EDIT: list of application groups
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapGroupFilterGroups "friends;family"
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapGroupFilterMode 0
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapGroupDisplayName cn
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapGroupFilterObjectclass groupOfUniqueNames
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapGroupMemberAssocAttr uniqueMember
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapEmailAttribute "mail"
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapLoginFilterEmail 0
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapLoginFilterUsername 1
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapMatchingRuleInChainState unknown
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapNestedGroups 0
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapPagingSize 500
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapTLS 0
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUserAvatarRule default
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUserDisplayName displayname
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUserFilterMode 1
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUuidGroupAttribute auto
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ ldap:set-config s01 ldapUuidUserAttribute auto

# Setup OIDC
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ app:install user_oidc
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ app:enable user_oidc
sudo docker exec --user www-data nextcloud-aio-nextcloud php occ user_oidc:provider Authelia \
--clientid="nextcloud" \
--clientsecret="secret" \
--discoveryuri="https://authelia.example.com/.well-known/openid-configuration" \
--scope="openid groups email profile" \
--mapping-uid="sub" \
--unique-uid=0 \
--no-interaction
```

Thank you

**Reply by hendrik1120** [reply]
Sure, no problem. After taking a brief look, I think changing the OIDC setting `--mapping-uid="sub" ` to `--mapping-uid="preferred_username" ` should already fix it.
I assume you want to have the username as the nextcloud uid instead of the ldap uuid or random uid right?
With ldap:show-remnants you can check all old ldap accounts and remove them.
You can disable oidc user creation with the following entry in config.php.
You are going to get an internal server error if oidc can't find the user in the ldap backend.
```php
 'user_oidc' => 
  array (
    'auto_provision' => false,
  ),
```
Feel free to ask if you have any questions

Edit: This might help in order to understand what kind of information authelia is providing:
https://www.authelia.com/integration/openid-connect/introduction/#openid

**Reply by Smithoo4** [reply]
Thank you for your help. I ended up wiping in everything and doing a clean installation to get things working, it was just a test environment.

Also, I appreciate the link, it was helpful. Is there a way to interrogate the openid scope values? I am thinking something like ldapsearch, I found this useful in understanding LDAP.

Cheers
